std-bits/base/src/main/java/org/enso/base/enso_cloud/TransientHTTPResponseCache.java

```diff
@@ -0,0 +1,419 @@
+    }
+  }
+
+  /**
+   * We define the TTL as the amount of time that the response should be considered fresh.
+   *
+   * <p>Define t0 as the time at which the content was generated on the origin server.
+   *
+   * <p>Define t1 as the time at which the current request was handled, either by the origin server
+   * or an intervening cache.
+   *
+   * <p>The 'Age' header, if present is (t1 - t0).
+   *
+   * <p>The 'max-age' value in the 'Cache-Control' header, if present, is the origin server's
+   * definition of how long the response should be considered fresh.
+   *
+   * <p>If 'max-age' and 'Age' are both present, we set TTL = max-age - Age. If only 'max-age' is
+   * present, we set TTL = max-age. If neither are present, we use a default.
+   */
+  private static int calculateTTL(HttpHeaders headers) {
+    Integer maxAge = getMaxAge(headers);
+    if (maxAge == null) {
+      return DEFAULT_TTL_SECONDS;
+    } else {
+      int age = headers.firstValue("age").map(Integer::parseInt).orElse(0);
+      return maxAge - age;
+    }
+  }
+
+  private static Integer getMaxAge(HttpHeaders headers) {
+    var cacheControlMaybe = headers.firstValue("cache-control");
+    Integer maxAge = null;
+    if (cacheControlMaybe.isPresent()) {
+      var cacheControl = cacheControlMaybe.get();
+      var cacheControlEntries = cacheControl.split(",");
+      for (var entry : cacheControlEntries) {
+        if (entry.trim().toLowerCase().startsWith("max-age")) {
+          var maxAgeBinding = entry.split("=");
+          if (maxAgeBinding.length > 1) {
+            maxAge = Integer.parseInt(maxAgeBinding[1]);
+          }
+          break;
+        }
+      }
+    }
+    return maxAge;
+  }
+
+  /**
+   * Sorts the header by header name, so we don't depend on header order. Multiple-valued headers
+   * might hash differently, but it's a rare case.
+   */
+  private static String makeHashKey(URI resolvedURI, List<Pair<String, String>> resolvedHeaders) {
+    try {
+      MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
+      messageDigest.update(resolvedURI.toString().getBytes());
+
+      var sortedHeaders = resolvedHeaders.stream().sorted(headerNameComparator).toList();
+      for (Pair<String, String> resolvedHeader : sortedHeaders) {
+        messageDigest.update(resolvedHeader.getLeft().getBytes());
+        messageDigest.update(resolvedHeader.getRight().getBytes());
+      }
+      return HexFormat.of().formatHex(messageDigest.digest());
+    } catch (NoSuchAlgorithmException ex) {
+      throw raise(RuntimeException.class, ex);
+    }
+  }
+
+  @FunctionalInterface
+  public interface RequestMaker {
+    EnsoHttpResponse run() throws IOException, InterruptedException;
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <E extends Exception> E raise(Class<E> type, Exception ex) throws E {
+    throw (E) ex;
+  }
+
+  private record CacheEntry(
+      URI uri,
+      HttpHeaders headers,
+      String responseDataPath,
```

Comment:
Why is this a `String` and not a more appropriate type like `File`?

I see that you are calling `File.getAbsolutePath` when storing this and then recovering with `new File(...)`. Wouldn't it be much simpler to just be passing this `File` around?

Reply:
Done